src/nvim/api/vim.c

```diff
@@ -663,6 +663,110 @@ Array nvim_get_api_info(uint64_t channel_id)
   return rv;
 }
 
+/// Call many api functions atomically
+///
+/// This has two main usages: Firstly, to perform several requests from an
+/// async context atomically, i.e. without processing requests from other rpc
+/// clients or redrawing or allowing user interaction in between. Note that api
+/// methods that could fire autocommands or do event processing still might do
+/// so. For instance invoking the :sleep command might call timer callbacks.
+/// Secondly, it can be used to reduce rpc overhead (roundtrips) when doing
+/// many requests in sequence.
+///
+/// @param calls an array of calls, where each call is described by an array
+/// with at least three elements: the request name, an array of arguments, and
+/// a boolean indicating if the return value should be sent back.
+///
+/// @return an array with three elements. The first is the number of calls that
+/// where performed without errors. If it is the same as the length of "calls",
+/// all calls were successful, otherwise it is the number of the last attempted
+/// call resulting in an error. The second is an array of return values, for
+/// calls when the third boolean flag was set to true. The third is NIL if all
+/// calls succeeded. If a call resulted in an error, it is a two-element array
+/// with the error code and message.
+Array nvim_multi_request(uint64_t channel_id, Array calls, Error *err)
+  FUNC_API_NOEVAL
+{
+  Array rv = ARRAY_DICT_INIT;
+  Array results = ARRAY_DICT_INIT;
+  // make place for the number of successful requests
+  Error nested_error = ERROR_INIT;
+
+  size_t i;  // also used for freeing the variables
+  for (i = 0; i < calls.size; i++) {
+    if (calls.items[i].type != kObjectTypeArray) {
+      api_set_error(err,
+                    Validation,
+                    _("All items in calls array must be arrays"));
+      goto end;
+    }
+    Array call = calls.items[i].data.array;
+    if (call.size != 3) {
+      api_set_error(err,
+                    Validation,
+                    _("All items in calls array must be arrays of size 3"));
+      goto end;
+    }
+
+    if (call.items[0].type != kObjectTypeString) {
+      api_set_error(err,
+                    Validation,
+                    _("name must be String"));
+      goto end;
+    }
+    String name = call.items[0].data.string;
```

Comment:
They could, but to me matching a field from a discriminated union is one operation, which happens to be two steps in language like C, thus best kept together.